ocaml/networkd/test/test_network_device_order.ml

```diff
@@ -0,0 +1,484 @@
+
+let pci_addr5 = pci_addr4
+
+let mac_addr0 = Macaddr.of_string "ec:f4:bb:e6:d7:b8" |> Result.get_ok
+
+let mac_addr1 = Macaddr.of_string "ec:f4:bb:e6:d7:b9" |> Result.get_ok
+
+let mac_addr2 = Macaddr.of_string "ec:f4:bb:e6:d7:ba" |> Result.get_ok
+
+let mac_addr3 = Macaddr.of_string "ec:f4:bb:e6:d7:bb" |> Result.get_ok
+
+let mac_addr4 = Macaddr.of_string "00:02:c9:ed:fd:f0" |> Result.get_ok
+
+let mac_addr5 = Macaddr.of_string "00:02:c9:ed:fd:f1" |> Result.get_ok
+
+let name0 = "eno1"
+
+let name1 = "eno2"
+
+let name2 = "eno3"
+
+let name3 = "eno4"
+
+let name4 = "enp5s0"
+
+let name5 = "enp5s0d1"
+
+let last_order =
+  [
+    {name= name0; pci= pci_addr0; mac= mac_addr0; position= 0; present= true}
+  ; {name= name1; pci= pci_addr1; mac= mac_addr1; position= 1; present= true}
+  ; {name= name2; pci= pci_addr2; mac= mac_addr2; position= 2; present= true}
+  ; {name= name3; pci= pci_addr3; mac= mac_addr3; position= 3; present= true}
+  ; {name= name4; pci= pci_addr4; mac= mac_addr4; position= 4; present= true}
+  ; {name= name5; pci= pci_addr5; mac= mac_addr5; position= 5; present= true}
+  ]
+
+let currents =
+  let open Dev in
+  [
+    {
+      name= name0
+    ; pci= pci_addr0
+    ; mac= mac_addr0
+    ; bios_eth_order= 0
+    ; multi_nic= false
+    }
+  ; {
+      name= name1
+    ; pci= pci_addr1
+    ; mac= mac_addr1
+    ; bios_eth_order= 1
+    ; multi_nic= false
+    }
+  ; {
+      name= name2
+    ; pci= pci_addr2
+    ; mac= mac_addr2
+    ; bios_eth_order= 2
+    ; multi_nic= false
+    }
+  ; {
+      name= name3
+    ; pci= pci_addr3
+    ; mac= mac_addr3
+    ; bios_eth_order= 3
+    ; multi_nic= false
+    }
+  ; {
+      name= name4
+    ; pci= pci_addr4
+    ; mac= mac_addr4
+    ; bios_eth_order= 5
+    ; multi_nic= false
+    }
+  ; {
+      name= name5
+    ; pci= pci_addr5
+    ; mac= mac_addr5
+    ; bios_eth_order= 4
+    ; multi_nic= false
+    }
+  ]
+
+let pos_of_mac mac order =
+  match List.find_opt (fun dev -> dev.mac = mac) order with
+  | Some {position; _} ->
+      position
+  | _ ->
+      -1
+
+let present_of_mac mac order =
+  match List.find_opt (fun dev -> dev.mac = mac) order with
+  | Some {present; _} ->
+      present
+  | _ ->
+      failwith "Can't find the device!"
+
+let test_default () =
+  let order = sort' ~currents ~rules:[] ~last_order:[] in
+  Alcotest.(check bool) "is Ok" true (Result.is_ok order) ;
+  let order = Result.get_ok order in
+  Alcotest.(check int) "6 devices in the order" 6 (List.length order) ;
+  Alcotest.(check int) "position assigned" 0 (pos_of_mac mac_addr0 order) ;
+  Alcotest.(check int) "position assigned" 1 (pos_of_mac mac_addr1 order) ;
+  Alcotest.(check int) "position assigned" 2 (pos_of_mac mac_addr2 order) ;
+  Alcotest.(check int) "position assigned" 3 (pos_of_mac mac_addr3 order) ;
+  Alcotest.(check int) "position assigned" 5 (pos_of_mac mac_addr4 order) ;
+  Alcotest.(check int) "position assigned" 4 (pos_of_mac mac_addr5 order) ;
```

Comment:
Why is this the order? mac_addr5 seems lower than 4  (The PCI address is the same for both)

Reply:
It's because of the bios_eth_order